src/moscot/utils/tagged_array.py

```diff
@@ -159,3 +172,205 @@ def is_kernel(self) -> bool:
+class DistributionContainer:
+    """Data container for OT problems involving more than two distributions.
+
+    TODO
+
+    Parameters
+    ----------
+    xy
+        Distribution living in a shared space.
+    xx
+        Distribution living in an incomparable space.
+    a
+        Marginals when used as source distribution.
+    b
+        Marginals when used as target distribution.
+    conditions
+        Conditions for the distributions.
+    cost_xy
+        Cost function when in the shared space.
+    cost_xx
+        Cost function in the incomparable space.
+    """
+
+    xy: Optional[ArrayLike]
+    xx: Optional[ArrayLike]
+    a: ArrayLike
+    b: ArrayLike
+    conditions: Optional[ArrayLike]
+    cost_xy: OttCostFn_t
+    cost_xx: OttCostFn_t
+
+    @staticmethod
+    def _extract_data(
+        adata: AnnData,
+        *,
+        attr: Literal["X", "obs", "obsp", "obsm", "var", "varm", "layers", "uns"],
+        key: Optional[str] = None,
+    ) -> ArrayLike:
+        modifier = f"adata.{attr}" if key is None else f"adata.{attr}[{key!r}]"
+        data = getattr(adata, attr)
+
+        try:
+            if key is not None:
+                data = data[key]
+        except KeyError:
+            raise KeyError(f"Unable to fetch data from `{modifier}`.") from None
+        except IndexError:
+            raise IndexError(f"Unable to fetch data from `{modifier}`.") from None
+
+        if sp.issparse(data):
+            logger.warning(f"Densifying data in `{modifier}`")
+            data = data.A
+        if data.ndim != 2:
+            raise ValueError(f"Expected `{modifier}` to have `2` dimensions, found `{data.ndim}`.")
+
+        return data
+
+    @staticmethod
+    def _verify_input(
+        xy_attr: Optional[Literal["X", "obsp", "obsm", "layers", "uns"]],
+        xy_key: Optional[str],
+        xx_attr: Optional[Literal["X", "obsp", "obsm", "layers", "uns"]],
+        xx_key: Optional[str],
+        conditions_attr: Optional[Literal["obs", "var", "obsm", "varm", "layers", "uns"]],
+        conditions_key: Optional[str],
+    ) -> Tuple[bool, bool, bool]:
+        if (xy_attr is None and xy_key is not None) or (xy_attr is not None and xy_key is None):
+            raise ValueError(r"Either both `xy_attr` and `xy_key` must be `None` or none of them.")
+        if (xx_attr is None and xx_key is not None) or (xx_attr is not None and xx_key is None):
+            raise ValueError(r"Either both `xy_attr` and `xy_key` must be `None` or none of them.")
+        if (conditions_attr is None and conditions_key is not None) or (
+            conditions_attr is not None and conditions_key is None
+        ):
+            raise ValueError(r"Either both `conditions_attr` and `conditions_key` must be `None` or none of them.")
+        return xy_attr is not None, xx_attr is not None, conditions_attr is not None
+
+    @classmethod
+    def from_adata(
+        cls,
+        adata: AnnData,
+        a: ArrayLike,
```

Comment:
Why not use uniform marginals as defaults?

Reply:
because we create them before already, so not necessary. Still, we could have a default value here, wdyt